File: DeckOfCards-iOS/DeckOfCardsTests/Classes/DeckTests.m
//
//  DeckTests.m
//  DeckOfCards
//
//  Created by Andrea Prearo on 11/20/13.
//  Copyright (c) 2013 Andrea Prearo. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "Deck.h"
#import "Card.h"

@interface DeckTests : XCTestCase

@property (nonatomic, retain) Deck *deck;

@end

@implementation DeckTests : XCTestCase

@synthesize deck = _deck;

- (void)setUp
{
	[super setUp];
	
	_deck = [Deck create];
}

- (void)tearDown
{
	[super tearDown];
}

- (void)testShuffle
{
	NSArray *cards1 = [self.deck.remainingCards copy];
	[self.deck shuffle];
	NSArray *cards2 = [self.deck.remainingCards copy];
	XCTAssertFalse([cards1 isEqualToArray:cards2], @"Deck has not been shuffled");
}

- (void)testDealOne
{
	NSUInteger len0 = self.deck.size;
	Card *card1 = [self.deck dealOne];
	NSUInteger len1 = self.deck.size;
	Card *card2 = [self.deck dealOne];
	NSUInteger len2 = self.deck.size;
	XCTAssertFalse([card1 isEqual:card2], @"Card has already been dealt");
	XCTAssertTrue(len0 - len1 == 1, @"Dealt card has not been removed from the deck");
	XCTAssertTrue(len1 - len2 == 1, @"Dealt card has not been removed from the deck");
}

- (void)testNoMoreCards
{
	NSUInteger count = self.deck.size;
	for (NSUInteger i = 0; i < count; i++) {
		[self.deck dealOne];
	}
	XCTAssertThrowsSpecificNamed([self.deck dealOne], NSException, @"InvalidOperationException",
															 @"All cards have been dealt");
}

- (void)testSameCard
{
	Card *card1 = [self.deck getCard:0];
	Card *card2 = [self.deck getCard:0];
	XCTAssertTrue([card1 isEqual:card2], @"Cards are different");
	Card *card3 = [Card createWithSuit:card1.suit andRank:card1.rank];
	XCTAssertTrue([card1 isEqual:card3], @"Cards are different");
}

- (void)testDifferentCard
{
	Card *card1 = [self.deck getCard:0];
	Card *card2 = [self.deck getCard:1];
	XCTAssertFalse([card1 isEqual:card2], @"Cards are the same");
}

@end

File: DeckOfCards-Swift/DeckOfCardsTests/Classes/DeckTests.swift
//
//  DeckTests.swift
//  DeckOfCards
//
//  Created by Andrea Prearo on 12/30/14.
//  Copyright (c) 2014 Andrea Prearo. All rights reserved.
//

import UIKit
import XCTest

class DeckTests: XCTestCase {
    var deck : Deck?

    override func setUp() {
        deck = Deck()
        super.setUp()
    }
    
    override func tearDown() {
        super.tearDown()
    }

    func testShuffle() {
        let cards1 = deck?.remainingCards
        deck?.shuffle()
        let cards2 = deck?.remainingCards
        XCTAssertNotEqual(cards1!, cards2!, "Deck has not been shuffled")
    }

    func testDealOne() {
        let len0 = deck?.size
        let card1 = deck?.dealOne()
        let len1 = deck?.size
        let card2 = deck?.dealOne()
        let len2 = deck?.size
        XCTAssertNotEqual(card1, card2, "Deck has not been shuffled")
        XCTAssertEqual(len0! - len1!, 1, "Dealt card has not been removed from the deck")
        XCTAssertEqual(len1! - len2!, 1, "Dealt card has not been removed from the deck")
    }

    func testNoMoreCards() {
        for _ in 1 ... deck!.size {
            _ = deck?.dealOne()
        }
        let card = deck?.dealOne()
        XCTAssertNil(card, "There are still cards in the deck")
    }

    func testSameCard() {
        let card1 = deck?.getCard(1)
        let card2 = deck?.getCard(1)
        XCTAssertEqual(card1, card2, "Cards are different")
        let card3 = Card(suit: card1?.suit, rank: card1?.rank)
        XCTAssertEqual(card1, card3, "Cards are different")
    }
    
    func testDifferentCard() {
        let card1 = deck?.getCard(1)
        let card2 = deck?.getCard(2)
        XCTAssertNotEqual(card1, card2, "Cards are the same")
    }
}
